row.py

```diff
@@ -988,9 +980,131 @@ def join_spreadsheet_info(filtered_results_file, out_dir):
+    """
+
+    #: silence pandas SettingWithCopyWarning
+    pd.options.mode.chained_assignment = None
+
+    #: set up variables
+    out_dir = Path(out_dir)
+    out_dir.mkdir(parents=True, exist_ok=True)
+
+    working_df = pd.read_csv(previous_results_file)
+
+    #: ensure the 'text' field is read as a string
+    working_df["text"] = working_df["text"].astype(str)
+
+    #: add a 'keep' field, initialized as 'yes', but be used to mark rows to remove with 'no'
+    working_df["keep"] = "yes"
+
+    #: upper-case all letters
+    working_df["text"] = working_df["text"].str.upper()
+
+    #: remove periods, parenthesis, double colons, and slashes
+    logging.info("Removing periods, parenthesis, double colons, and slashes")
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace(".", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("(", "").replace("(", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace(")", "").replace(")", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("/", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("\\", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("::", ":").strip(), axis=1)
+
+    #: remove leading colons
+    logging.info("Removing leading colons")
+    working_df["text"] = working_df["text"].apply(lambda x: x.lstrip(":"))
+
+    #: replace euro symbol with 'E'
+    logging.info("Replacing euro symbol with 'E'")
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("€", "E").strip(), axis=1)
+
+    #: replaced accented characters with non-accented characters
+    logging.info("Replacing accented characters with non-accented characters")
+    working_df["text"] = working_df.apply(
+        lambda r: "".join(c for c in unicodedata.normalize("NFD", r["text"]) if unicodedata.category(c) != "Mn"), axis=1
+    )
+
+    #: flag parcels that contains special characters
+    working_df["special_char"] = "pass"
+    mask = working_df["text"].str.contains(r"[^\w:]") == True
+    working_df.loc[mask, "keep"] = "no"
+    working_df.loc[mask, "special_char"] = "fail"
+    logging.info("Number parcels with special characters flagged: %i", mask.value_counts()[1])
+
+    #: flag parcels that start with a letter or non-digit, should start with a number
+    working_df["nondigit_start"] = "pass"
+    mask = ~working_df["text"].str.contains(r"^\d") == True
+    working_df.loc[mask, "keep"] = "no"
+    working_df.loc[mask, "nondigit_start"] = "fail"
+    logging.info("Number parcels starting with a non-digit flagged: %i", mask.value_counts()[1])
+
+    #: flag parcels with a ':P' pattern; permits, not relevant - but PUE is okay!
+    working_df["permit"] = "pass"
+    mask = working_df["text"].str.contains(r"(?!:PUE):P") == True
```

Comment:
**question**: would `:P$` be a simpler regex for this?

do we need to be ignoring case for any of these alpha matches?

Reply:
I call `working_df["text"].str.upper()` early on because all letters should be upper-case, so the rules focus on upper-case.
`(?!:PUE):P` is slightly than `:P$` different because I don't want to require it to be at the end of a string.  I'll add an example.